src/Symfony/Component/OptionsResolver/Tests/OptionsResolverTest.php

```diff
@@ -778,6 +778,33 @@ public function testSetAllowedTypesFailsIfUnknownOption()
         $this->resolver->setAllowedTypes('foo', 'string');
     }
 
+    public function testResolveTypedWithUnion()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', 'string|int');
+        $options = $this->resolver->resolve(['foo' => 1]);
+
+        $this->assertSame(['foo' => 1], $options);
+    }
+
+    public function testResolveTypedWithUnionUselessParentheses()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', '(string)|(int)');
+        $options = $this->resolver->resolve(['foo' => 1]);
+
+        $this->assertSame(['foo' => 1], $options);
+    }
+
+    public function testResolveTypedWithUnionUselessParentheses2()
+    {
+        $this->resolver->setDefined('foo');
+        $this->resolver->setAllowedTypes('foo', '((string)|(int))');
```

Comment:
If useless, I'd prefer to throw an exception (invalid type syntax) instead, as this could be a case where you might be missing the `[]` at the end. I believe an exception that forces you to remove the unnecessary parentheses would be more helpful than allowing them for no reason.

Reply:
What do you consider as useless parentheses ?
- (A) ?
- (A|B) ? But since the called was made as recursive it might come from `(A|B)[]`
- (A|B)|(C|D) ? But this doesn't seems easy to report this as useless since
  - It's split as `(A|B)` and `(C|D)` and then we're in the previous case

Since the call is recursive it doesn't seems easy to report useless parenthesis or maybe it will require an extra param to know if we just removed `[]`.

Also, since the validation is made in Runtime (when calling `resolve`) it seems not as useful as it would have been when calling `setAllowedTypes`...

Reply:
Also, there is so far no similar validation about allowedTypes passed.
- `string[` is not reported even if it's obviously a mistake
- `['string', 'string']` is not reported
- `strin[]` is not reported

I think introducing an InvalidType exceptions for those syntax and useless parenthesis might be better in another PR.

Reply:
Agree, but let’s not add test cases for them though

Reply:
Done